File: 438/solution.go
func findAnagrams(s string, p string) []int {
	pv := make([]int, 26)
	sv := make([]int, 26)
	ans := make([]int, 0)

	if len(s) < len(p) {
		return ans
	}

	for i := 0; i < len(p); i++ {
		sv[charToInt(s[i])]++
		pv[charToInt(p[i])]++
	}

	if isSliceEqual(pv, sv) {
		ans = append(ans, 0)
	}

	for i := len(p); i < len(s); i++ {
		sv[charToInt(s[i])]++
		sv[charToInt(s[i-len(p)])]--

		if isSliceEqual(pv, sv) {
			ans = append(ans, i-len(p)+1)
		}
	}

	return ans
}

func isSliceEqual(a []int, b []int) bool {
	for idx, val := range a {
		if val != b[idx] {
			return false
		}
	}
	return true
}

func charToInt(char uint8) int {
	return int(char - 'a')
}

File: 438/solution.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> sv(26, 0), pv(26, 0), ans;
        
        if (s.size() < p.size()) return ans;
        
        for (int i=0; i<p.size(); i++) {
            sv[charToInt(s[i])]++;
            pv[charToInt(p[i])]++;
        }
        
        if (sv == pv) ans.push_back(0);
        
        
        for (int i=p.size(); i<s.size(); i++) {
            
            // New element has come
            sv[charToInt(s[i])]++;
            
            // Remove the old first element
            sv[charToInt(s[i-p.size()])]--;
            
            if (pv == sv) {
                ans.push_back(i-p.size()+1);
            }
        }
        
        return ans;
    }
    
    int charToInt(char c) {
        return c - 'a';
    }
};